### PR TITLE
feat(dropdown) Add keynav support (fix for #1228)

### PR DESCRIPTION
**Pull request by bleggett**
Adds a new option flag to the dropdown class that enables the use of arrow keys to navigate dropdown list elements, like non-angular bootstrap does. Added tests and updated doc with example. Fix for [#1228] (https://github.com/angular-ui/bootstrap/issues/1228) and based off of [#3212] (https://github.com/angular-ui/bootstrap/issues/3212)

### CODE REVIEW

**Comment by rvanbaalen** [comment]
Please squash your commits.

**Comment by bleggett** [comment]
Squashed.

File: src/dropdown/dropdown.js [hunk 1]
```diff
@@ -112,6 +120,40 @@ angular.module('ui.bootstrap.dropdown', ['ui.bootstrap.position'])
...
+      case (40): {
+        if ( !angular.isNumber(self.selectedOption)) {
+          self.selectedOption = 0;
+        } else {
+          self.selectedOption = (self.selectedOption === elems.length -1 ?
+                                 self.selectedOption :
+                                 self.selectedOption+1);
+        }
+      }
+      break;
```
Comment (wesleycho) [inline src/dropdown/dropdown.js]: Shouldn't this be inside the `case` statement?

File: src/dropdown/dropdown.js [hunk 2]
```diff
@@ -112,6 +120,40 @@ angular.module('ui.bootstrap.dropdown', ['ui.bootstrap.position'])
...
+      case (38): {
+        if ( !angular.isNumber(self.selectedOption)) {
+          return;
+        } else {
+          self.selectedOption = (self.selectedOption === 0 ?
+                                 0 :
+                                 self.selectedOption-1);
+        }
+      }
+      break;
```
Comment (wesleycho) [inline src/dropdown/dropdown.js]: Ditto

File: src/dropdown/test/dropdown.spec.js [hunk 3]
```diff
@@ -213,7 +213,7 @@ describe('dropdownToggle', function() {
       });
 
       return $compile('<li dropdown><a href dropdown-toggle></a>' +
-        '<ul><li><a href="#something">Hello</a></li></ul></li>')($rootScope);
+                      '<ul><li><a href="#something">Hello</a></li></ul></li>')($rootScope);
```
Comment (wesleycho) [inline src/dropdown/test/dropdown.spec.js]: Revert this change

File: src/dropdown/test/dropdown.spec.js [hunk 4]
```diff
@@ -350,8 +350,8 @@ describe('dropdownToggle', function() {
   describe('`auto-close` option', function() {
     function dropdown(autoClose) {
       return $compile('<li dropdown ' +
-        (autoClose === void 0 ? '' : 'auto-close="'+autoClose+'"') +
-        '><a href dropdown-toggle></a><ul><li><a href>Hello</a></li></ul></li>')($rootScope);
+                      (autoClose === void 0 ? '' : 'auto-close="'+autoClose+'"') +
+                      '><a href dropdown-toggle></a><ul><li><a href>Hello</a></li></ul></li>')($rootScope);
```
Comment (wesleycho) [inline src/dropdown/test/dropdown.spec.js]: Revert the indentation change in these two lines

**Comment by wesleycho** [comment]
These are mostly code style issues. @rvanbaalen if you think it's worth merging and fixing these issues before pushing up, this PR LGTM.

File: src/dropdown/dropdown.js (hunk 1 above)
Comment (rvanbaalen) [inline src/dropdown/dropdown.js]: You're saying 

```js
var elems = self.dropdownMenu ? //If append to body is used.
      (angular.element(self.dropdownMenu).find('a')) :
     (angular.element(self.$element).find('ul').eq(0).find('a'));
```

should be copied to each of these `case`'s?

Comment (wesleycho) [inline src/dropdown/dropdown.js]: I mean the `break`

File: src/dropdown/dropdown.js [hunk 5]
```diff
@@ -112,6 +120,40 @@ angular.module('ui.bootstrap.dropdown', ['ui.bootstrap.position'])
...
+      (angular.element(self.$element).find('ul').eq(0).find('a'));
+
+    switch (keyCode) {
+      case (40): {
+        if ( !angular.isNumber(self.selectedOption)) {
+          self.selectedOption = 0;
+        } else {
+          self.selectedOption = (self.selectedOption === elems.length -1 ?
+                                 self.selectedOption :
+                                 self.selectedOption+1);
```
Comment (wesleycho) [inline src/dropdown/dropdown.js]: Add spaces before and after the `+`.

Also make the indentation only one level deep, and not align to the `self` in the first line.

File: src/dropdown/dropdown.js [hunk 6]
```diff
@@ -112,6 +120,40 @@ angular.module('ui.bootstrap.dropdown', ['ui.bootstrap.position'])
...
+        }
+      }
+      break;
+      case (38): {
+        if ( !angular.isNumber(self.selectedOption)) {
+          return;
+        } else {
+          self.selectedOption = (self.selectedOption === 0 ?
+                                 0 :
+                                 self.selectedOption-1);
```
Comment (wesleycho) [inline src/dropdown/dropdown.js]: Add spaces before and after the `-`.

Also change indentation to only be one level deep, not align to `self` from the first line.

**Comment by RobJacobs** [comment]
My only concern is with setting focus() on the menu item elements, I'm not sure if this would disrupt the tab flow once the dropdown is closed?  The typeahead has similar functionality but is uses an activeIndex with:

   ng-class="{active: isActive($index) }"

to decorate the element.  Then listens for enter to trigger selection.

File: src/dropdown/dropdown.js (hunk 1 above)
Comment (rvanbaalen) [inline src/dropdown/dropdown.js]: Of course. 

The official C99 specification says the following about the break statement:
> A break statement terminates execution of the smallest enclosing switch or iteration statement.

So I would vote to put these breaks inside the curly brackets as well.

**Comment by rvanbaalen** [comment]
@bleggett Can you incorporate these suggestions?

**Comment by bleggett** [comment]
@rvanbaalen Yep, will do. Will look at what typeahead is doing like @RobJacobs suggested as well.

**Comment by wesleycho** [comment]
Can you squash your commits again? In the future, if you need to sync up with remote master, you can do `git pull --rebase origin master` (assuming origin is pointing to this repository)

**Comment by bleggett** [comment]
@wesleycho Yeah, resquashed. Still getting familiar with git, so forgive the rookie mistakes.

Regarding @RobJacobs suggestion: typeahead creates the menu element dynamically, and so can add attrs to it (like the select and activeIndex) relatively easily.

Dropdown relies on a predefined, user-defined menu element, and I'm less sure about tacking attrs onto that blindly, or expecting users to add this stuff themselves.

Am I way off here?

**Comment by RobJacobs** [comment]
@bleggett Thanks for looking at that, I agree with your assessment.
